lib/promscrape/discovery/consulagent/watch.go

```diff
@@ -0,0 +1,297 @@
+	serviceWatchersStopped = metrics.NewCounter("vm_promscrape_discovery_consulagent_service_watchers_stopped_total")
+	serviceWatchersCount   = metrics.NewCounter("vm_promscrape_discovery_consulagent_service_watchers")
+)
+
+// getServiceNames obtains serviceNames via request to Consul Agent.
+//
+// It returns an empty serviceNames list if response contains the same index.
+func (cw *consulAgentWatcher) getServiceNames() ([]string, error) {
+	path := "/v1/agent/services" + cw.servicesQueryArgs
+	data, err := cw.client.GetAPIResponse(path)
+	if err != nil {
+		return nil, fmt.Errorf("cannot perform Consul Agent API request at %q: %w", path, err)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+	var m map[string]consul.Service
+	if err := json.Unmarshal(data, &m); err != nil {
+		return nil, fmt.Errorf("cannot parse response from %q: %w; data=%q", path, err, data)
+	}
+	serviceNames := make([]string, 0, len(m))
+	for serviceName, service := range m {
+		if service.Datacenter != cw.watchDatacenter {
+			continue
+		}
+		if !shouldCollectServiceByName(cw.watchServices, serviceName) {
+			continue
+		}
+		serviceNames = append(serviceNames, serviceName)
+	}
+	return serviceNames, nil
+}
+
+// watchForServiceNodesUpdates watches for Consul serviceNode changes for the given serviceName.
+//
+// watchForServiceNodesUpdates calls initWG.Done() once the initialization is complete and the first discovery iteration is done.
+func (sw *serviceWatcher) watchForServiceNodesUpdates(cw *consulAgentWatcher, initWG *sync.WaitGroup) {
+	apiServer := cw.client.APIServer()
+	path := "/v1/agent/health/service/name/" + sw.serviceName
+	f := func() {
+		data, err := cw.client.GetAPIResponseWithParamsCtx(sw.requestCtx, path, nil, nil)
+		if err != nil {
+			if !errors.Is(err, context.Canceled) {
+				logger.Errorf("cannot obtain Consul Agent serviceNodes for serviceName=%q from %q: %s", sw.serviceName, apiServer, err)
+			}
+			return
+		}
+		sns, err := consul.ParseServiceNodes(data)
+		if err != nil {
+			logger.Errorf("cannot parse Consul Agent serviceNodes response for serviceName=%q from %q: %s", sw.serviceName, apiServer, err)
+			return
+		}
+
+		cw.servicesLock.Lock()
+		sw.serviceNodes = sns
+		cw.servicesLock.Unlock()
+	}
+
+	f()
+	// Notify caller that initialization is complete
+	initWG.Done()
+
+	ticker := time.NewTicker(getCheckInterval())
+	defer ticker.Stop()
+	stopCh := sw.requestCtx.Done()
+	for {
+		select {
+		case <-ticker.C:
+			f()
+		case <-stopCh:
+			return
+		}
+	}
+}
+
+// getServiceNodesSnapshot returns a snapshot of discovered ServiceNodes.
+func (cw *consulAgentWatcher) getServiceNodesSnapshot() map[string][]consul.ServiceNode {
+	cw.servicesLock.Lock()
+	sns := make(map[string][]consul.ServiceNode, len(cw.services))
+	for svc, sw := range cw.services {
+		sns[svc] = sw.serviceNodes
+	}
+	cw.servicesLock.Unlock()
+	return sns
+}
+
+func shouldCollectServiceByName(filterServices []string, serviceName string) bool {
```

Comment:
Should it use function from consul discovery as functions seems to be the same?

Reply:
sure, done